### ISSUE TITLE
Change semantics of `return` to be an early return

### ISSUE THREAD

**Issue by b-studios** [effekt-lang/effekt#404]
Right now `return` is a keyword but only used to embed expressions into statements. This is fine for a core representation like System C, but not appropriate for source.

That is, right now
```
def foo() = {
  val x = if (true) return 42 else 0;
  x + 1
}
```
currently returns `43` and not `42` (which might be confusing to people).

**Comment by b-studios** [comment]
yes, main point is that it is very confusing the way it is now. For an IR `return` is fine, for the source language, the way it currently is, is not.

Currently `Control` as in the stdlib is not good enough for boundary/break, since it does not take arguments. I guess we should change this, then.

**Comment by jiribenes** [comment]
Agreed, parametrising `Control` sounds good, the biggest problem I can foresee is the following:

What should be the type of `foreach`?
```scala
// currently
def foreach[A](l: List[A]) { f: A => Unit / Control } : Unit

// Option 1: we can only use `do break(())`, i.e. `Control[Unit]`
// ... but what if I _want_ to return some value
def foreach[A](l: List[A]) { f: A => Unit / Control[Unit] } : Unit

// Option 2: we align the type of `break` with the return type
// ... but what if I do _not_ want to return a value? Perhaps this needs to return `Option[R]`?
// ... also we might have errors/warnings because we're ignoring a returned value?
def foreach[A, R](l: List[A]) { f: A => Unit / Control[R] } : R
// which is strictly more general than Option 1 and reduces to it if we break with `()`
```
Are there any other possibilities / designs?

**Comment by b-studios** [comment]
It needs to be `Control[Unit]` since in the case of not returning early we would otherwise need to make up an `R`.

**Comment by jiribenes** [comment]
I had a quick thought: If we use `Control[R]` as `boundary/break`, what should the semantics of `continue` be in the `boundary` handler?

**Comment by b-studios** [comment]
There are only two options, right? Either it aligns with break, or it restarts :)

**Comment by b-studios** [comment]
I just tried adding an argument to `break`, but it is really annoying in the cases, where `Control[Unit]` and we would like to call `label.break()`; here we need to write `label.break(())`, then.

**Comment by omentic** [comment]
Hi @b-studios & co - I'm coming back to Effekt after a little while and am hitting conceptual issues with `return`, now that handlers are overt functions. How might one return early out of the top-level function, instead of just the handler? Is the intended way to set overt blocks and mutate variables and break out?

This is my example code, which doesn't compile due to a type mismatch:
```scala
interface Yield[T] {
  def yield(value: T): Unit
}

def yield[T](value: T) = {
  do yield(value)
}

def find_map[T, U]() {self: => Unit / Yield[T]} {fn: T => Option[U]}: Option[U] = {
  try self() with Yield[T] {
    def yield(x) = {
      val res = fn(x)
      res match {
        // cannot merge different constructors: Option vs. Unit
        // so this `return` is semantically relevant to `def yield` rather than `def find_map`
        case Some(_) => return res
        case None() => resume(())
      }
    }
  }
  None()
}
```

(I think the behavior of `return` as described here might mean my normal `find` function, which compiles but hasn't been run yet, will accidentally always return `None()`?)
```scala
def find[T]() {self: => Unit / Yield[T]} {fn: T => Bool}: Option[T] = {
  try self() with Yield[T] {
    def yield(x) = {
      if (fn(x)) {
        return Some(x) // error: 
      }
      resume(())
    }
  }
  None()
}
```

**Comment by b-studios** [comment]
Currently `return` is more of a relic of our papers where we have

```
s ::= val x = s; s
    | return e
    | ...
```

that is, it just embeds expressions into statements, without any control flow transfer whatsoever.

In your example, where would you expect `return` to be delimited. I see two choices: `yield` and `find`.

**Comment by omentic** [comment]
Hmm, okay, I guess I should probably avoid it then.

The intention was for `return` to delimit `find`, though of course it would make a lot more sense to delimit `yield`. The code I've been translating into Effekt was written for handlers not being functions exactly (so `return` would delimit `find`). I think using mutable variables does what I was originally intending:
```scala
def find[T]() {self: => Unit / Yield[T]} {fn: T => Bool}: Option[T] = {
  var res = None()
  try self() with Yield[T] {
    def yield(x) = {
      if (fn(x)) {
        res = Some(x)
      } else {
        resume(())
      }
    }
  }
  res
}

def find_map[T, U]() {self: => Unit / Yield[T]} {fn: T => Option[U]}: Option[U] = {
  var res = None()
  try self() with Yield[T] {
    def yield(x) = {
      fn(x) match {
        case Some(value) => res = Some(value)
        case None() => resume(())
      }
    }
  }
  res
}
```

**Comment by b-studios** [comment]
I think in your specific example I would do the following:

```
def find_map[T, U]() {self: => Unit / Yield[T]} {fn: T => Option[U]}: Option[U] = {
  try { self(); None() } with Yield[T] {
    def yield(x) = {
      fn(x) match {
        case Some(value) => Some(value)
        case None() => resume(())
      }
    }
  }
}
```

**Comment by omentic** [comment]
Oh, putting more stuff in the `try` block is a good idea, thanks. 

Also, this code is for a port of Rust's `std::iter` design to Effekt! If you're interested I have it mostly done here: https://gist.github.com/omentic/58b553920b4c8402b04c2873901291df. (I still need to figure out how syntax around calling functions that take in blocks works, and figure out how to implement `next()` and `zip()`, and make sure that everything that's currently compiling is actually correct.)

I think it's really cool that you can implement what Rust needs very complicated types and structs and such for with almost total effect inference, and get laziness "for free" :smile:

**Comment by b-studios** [comment]
People really like early returns: https://x.com/tsoding/status/1887896895936159970?t=O2uq21hzfm0cjRUnrI9Tfw&s=19
